src/core/ddsc/src/dds_domain.c

```diff
@@ -378,28 +385,50 @@ dds_return_t dds_domain_resolve_type (dds_entity_t entity, unsigned char *type_i
     goto failed;
   }
 
-  if (type_identifier == NULL || type_identifier_sz != TYPEID_HASH_LENGTH || sertype == NULL)
-  {
-    rc = DDS_RETCODE_BAD_PARAMETER;
-    goto failed;
-  }
-
-  type_identifier_t type_id;
-  memcpy (&type_id.hash, type_identifier, TYPEID_HASH_LENGTH);
-  struct tl_meta *tlm = ddsi_tl_meta_lookup (&e->m_domain->gv, &type_id);
+  struct ddsi_domaingv *gv = &e->m_domain->gv;
+  memcpy (&type_id.hash, type_identifier, sizeof (type_id));
+  struct tl_meta *tlm = ddsi_tl_meta_lookup (gv, &type_id);
   if (tlm == NULL)
   {
-    rc = DDS_RETCODE_NOT_FOUND;
+    rc = DDS_RETCODE_PRECONDITION_NOT_MET;
     goto failed;
   }
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
   if (tlm->state == TL_META_RESOLVED)
-    *sertype = tlm->sertype;
+  {
+    *sertype = ddsi_sertype_ref (tlm->sertype);
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  }
   else
   {
-    // FIXME: resolve the type information
-    //ddsi_tl_request_type (&e->m_domain->gv, &type_id);
-    rc = DDS_RETCODE_NOT_FOUND;
-    goto failed;
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    ddsrt_mutex_lock (&gv->tl_resolved_lock);
```

Comment:
I've refactored this so that it's now only using the `gv->tl_admin_lock` mutex. After requesting the type and re-acquiring the lock, there is a test to check if the type is already resolved, as it may have been resolved after the lock was released in `ddsi_tl_request_type` (which is required because when writing the sample we shouldn't hold the lock). In that case there is no need to wait for `gv->tl_resolved_cond` to be triggered. 

In `ddsi_tl_request_type` there is a test `tlm->state == TL_META_NEW` to avoid duplicate requests for a type, so I think that makes it the 'lookup-and-request-if-unknown' function you mean?